merkle-tree/src/oneshot.rs

```diff
@@ -5,27 +5,199 @@ use simperby_common::MerkleProof;
 ///
 /// This is useful for per-block data such as transaction lists.
 pub struct OneshotMerkleTree {
-    // TODO
+    hash_list: Vec<Hash256>,
 }
 
 impl OneshotMerkleTree {
     pub const EMPTY_HASH: Hash256 = Hash256::zero();
 
-    pub fn create(_data: Vec<Hash256>) -> Self {
-        unimplemented!()
+    /// Creates a new OneshotMerkleTree from the given data.
+    pub fn create(data: Vec<Hash256>) -> Self {
+        OneshotMerkleTree { hash_list: data }
     }
 
     /// Creates a Merkle proof for a given data in the tree.
     ///
     /// Returns `None` if the data is not in the tree.
-    pub fn create_merkle_proof(&self, _key: Hash256) -> Option<MerkleProof> {
-        unimplemented!()
+    pub fn create_merkle_proof(&self, mut key: Hash256) -> Option<MerkleProof> {
+        if !self.hash_list.contains(&key) {
+            return None;
+        }
+
+        let mut merkle_proof: MerkleProof = MerkleProof { proof: Vec::new() };
+        let mut merkle_tree: Vec<Vec<Hash256>> = Self::merkle_tree(&self.hash_list);
+
+        merkle_tree.pop();
+        for level in merkle_tree {
+            for pair in level.chunks(2) {
+                if pair.contains(&key) {
+                    let index: u8 = if pair[0] == key { 1u8 } else { 0u8 };
+                    merkle_proof
+                        .proof
+                        .push((pair[index as usize].clone(), index));
+                    key = Self::hash_pair(pair);
+                }
+            }
+        }
+        Some(merkle_proof)
+    }
+
+    /// Creates a merkle tree from the given hash list.
+    ///
+    /// Merkle tree is returned in the form of Vec of Vec of Hash256.
+    ///
+    /// Given a hash list [1, 2, 3, 4], merkle tree will be built as below,
+    ///
+    /// ``` text
+    ///     7
+    ///   5   6
+    ///  1 2 3 4
+    /// ```
+    /// which is represented as [[1, 2, 3, 4], [5, 6], [7]].
+    fn merkle_tree(hash_list: &[Hash256]) -> Vec<Vec<Hash256>> {
+        let mut merkle_tree: Vec<Vec<Hash256>> = vec![hash_list.to_vec()];
+
+        while merkle_tree.last().unwrap().len() != 1 {
+            let mut upper_level_hash_list: Vec<Hash256> = Vec::new();
+
+            for pair in merkle_tree.last().unwrap().chunks(2) {
+                upper_level_hash_list.push(Self::hash_pair(pair));
+            }
+
+            merkle_tree.push(upper_level_hash_list);
+        }
+
+        merkle_tree
+    }
```

Comment:
Please remove meaningless newlines in the method body

Reply:
Fixed.